Enable std-database in native image

hubertp opened the pull request:
### Pull Request Description

`enso --run test/Table_Tests` passes without failures.
Used the opportunity to cleanup resource configs a bit and update them for each of standard libraries (not yet complete but almost).

Closes #12006 and #12018 

### Important Notes

We include `java.desktop` module now, which is unfortunate but necessary for `java.beans` and `java.awt`.
But we also use NI substitution which workarounds current `java.awt` [limitations](https://github.com/oracle/graal/issues/4124).

More importantly, Language Server's NI continues to work.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Enable std-database in native image

Not working ATM, breaks even BASE stdlib tests.
Culprit appears to be `sqlite-jdbc` client

hubertp committed:
Workaround for sqlite-jdbc NI properies bug

sqlite-jdbc's NI properties file defines
`--enable-url-protocols=jar`. That's experimental and breaks completely
our `HostClassloader`. We exclude their config but include their feature
to correctly extract shared native library.

hubertp committed:
fmt

hubertp committed:
Towards error-free Table_Tests

hubertp committed:
More reflect updates to make NI happy

hubertp committed:
Segfaults in NI

hubertp committed:
Pass all Table_Tests except for Excel_Spec

Excel_Spec leads to segfault, potentially related to reading/writing
files.

hubertp commented:
Now, `Excel_Spec`'s segfaults are preceded by either
```
org.apache.poi.poifs.filesystem.NotOLE2FileException: Invalid header signature; read 0x6C61657220746F6E, expected 0xE11AB1A1E011CFD0 - Your file appears not to be a valid OLE2 document
	at org.apache.poi.poifs.storage.HeaderBlock.<init>(HeaderBlock.java:149)
	at org.apache.poi.poifs.storage.HeaderBlock.<init>(HeaderBlock.java:115)
	at org.apache.poi.poifs.filesystem.POIFSFileSystem.<init>(POIFSFileSystem.java:270)
	at org.apache.poi.poifs.filesystem.POIFSFileSystem.<init>(POIFSFileSystem.java:183)
	at org.enso.table.excel.ExcelConnectionPool.openWorkbook(ExcelConnectionPool.java:309)
	at org.enso.table.excel.ExcelConnectionPool$ConnectionRecord.reopen(ExcelConnectionPool.java:278)
	at org.enso.table.excel.ExcelConnectionPool.openReadOnlyConnection(ExcelConnectionPool.java:72)
	at java.base@21.0.2/java.lang.reflect.Method.invoke(Method.java:580)
```
or
```
org.apache.poi.poifs.filesystem.OfficeXmlFileException: The supplied data appears to be in the Office 2007+ XML. You are calling the part of POI that deals with OLE2 Office Documents. You need to call a different part of POI to process this data (eg XSSF instead of HSSF)
	at org.apache.poi.poifs.storage.HeaderBlock.<init>(HeaderBlock.java:128)
	at org.apache.poi.poifs.storage.HeaderBlock.<init>(HeaderBlock.java:115)
	at org.apache.poi.poifs.filesystem.POIFSFileSystem.<init>(POIFSFileSystem.java:270)
	at org.apache.poi.poifs.filesystem.POIFSFileSystem.<init>(POIFSFileSystem.java:183)
	at org.enso.table.excel.ExcelConnectionPool.openWorkbook(ExcelConnectionPool.java:309)
	at org.enso.table.excel.ExcelConnectionPool$ConnectionRecord.reopen(ExcelConnectionPool.java:278)
	at org.enso.table.excel.ExcelConnectionPool.openReadOnlyConnection(ExcelConnectionPool.java:72)
	at org.enso.table.read.ExcelReader.withWorkbook(ExcelReader.java:266)
	at org.enso.table.read.ExcelReader.readRangeByName(ExcelReader.java:181)
	at java.base@21.0.2/java.lang.reflect.Method.invoke(Method.java:580)
```
. randomly. Which feels like this is still related to some configuration.

hubertp commented:
Further narrowed it down to [Excel .xls](https://github.com/enso-org/enso/blob/develop/test/Table_Tests/src/IO/Excel_Spec.enso#L207). `.xlsx` seems fine.

hubertp committed:
Make sure `awt` is included in NI resource configs

Apparently `poi` requires `java.awt` and it was segfaulting in a
mysterious way when it was missing.

hubertp commented:
Here was a hint:
```
Stacktrace for the failing thread 0x00005db97911f2c0 (A=AOT compiled, J=JIT compiled, D=deoptimized, i=inlined):
  A  SP 0x00007ffdbdf141c0 IP 0x00005db960fa710e size=80    [image code] com.oracle.svm.core.code.IsolateEnterStub.JNIFunctions_FatalError_2d6e7de51007501356372d87ae0cc671dc9d1a91(IsolateEnterStub.java:0)
  i  SP 0x00007ffdbdf15270 IP 0x00005db961062904 size=80    [image code] com.oracle.svm.core.jni.JNIOnLoadFunctionPointer.invoke(JNILibraryInitializer.java)
  A  SP 0x00007ffdbdf15270 IP 0x00005db961062904 size=80    [image code] com.oracle.svm.core.jni.JNILibraryInitializer.callOnLoadFunction(JNILibraryInitializer.java:71)
  A  SP 0x00007ffdbdf152c0 IP 0x00005db9610635cc size=32    [image code] com.oracle.svm.core.jni.JNILibraryInitializer.initialize(JNILibraryInitializer.java:132)
  A  SP 0x00007ffdbdf152e0 IP 0x00005db961039d55 size=160   [image code] com.oracle.svm.core.jdk.NativeLibrarySupport.addLibrary(NativeLibrarySupport.java:204)
  A  SP 0x00007ffdbdf15380 IP 0x00005db96103a7a3 size=48    [image code] com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibrary0(NativeLibrarySupport.java:160)
  A  SP 0x00007ffdbdf153b0 IP 0x00005db96103b125 size=176   [image code] com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryRelative(NativeLibrarySupport.java:123)
  i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.lang.ClassLoader.loadLibrary(ClassLoader.java:106)
  i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.lang.Runtime.loadLibrary0(Runtime.java:916)
  i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.lang.System.loadLibrary(System.java:2063)
  i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.awt.Toolkit$2.run(Toolkit.java:1384)
  A  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.awt.Toolkit$2.run(Toolkit.java:1382)
  A  SP 0x00007ffdbdf15470 IP 0x00005db962af378a size=64    [image code] java.security.AccessController.executePrivileged(AccessController.java:129)
  i  SP 0x00007ffdbdf154b0 IP 0x00005db9626f4cdd size=16    [image code] java.security.AccessController.doPrivileged(AccessController.java:319)
  A  SP 0x00007ffdbdf154b0 IP 0x00005db9626f4cdd size=16    [image code] java.awt.Toolkit.loadLibraries(Toolkit.java:1381)
  A  SP 0x00007ffdbdf154c0 IP 0x00005db9626f41f0 size=32    [image code] java.awt.Toolkit.initStatic(Toolkit.java:1419)
  A  SP 0x00007ffdbdf154e0 IP 0x00005db9626f18a8 size=16    [image code] java.awt.Toolkit.<clinit>(Toolkit.java:1393)
  A  SP 0x00007ffdbdf154f0 IP 0x00005db960eabb6e size=16    [image code] com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:407)
  A  SP 0x00007ffdbdf15500 IP 0x00005db960ea9d7d size=80    [image code] com.oracle.svm.core.classinitialization.ClassInitializationInfo.doInitialize(ClassInitializationInfo.java:323)
  A  SP 0x00007ffdbdf15550 IP 0x00005db960eaa6f5 size=1296  [image code] com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:272)
  A  SP 0x00007ffdbdf15a60 IP 0x00005db9626da9ad size=144   [image code] java.awt.Color.<clinit>(Color.java:277)
  A  SP 0x00007ffdbdf15af0 IP 0x00005db960eabb6e size=16    [image code] com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:407)
  A  SP 0x00007ffdbdf15b00 IP 0x00005db960ea9d7d size=80    [image code] com.oracle.svm.core.classinitialization.ClassInitializationInfo.doInitialize(ClassInitializationInfo.java:323)
  A  SP 0x00007ffdbdf15b50 IP 0x00005db960eaa6f5 size=1296  [image code] com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:272)
  i  SP 0x00007ffdbdf16060 IP 0x00005db963852335 size=1216  [image code] org.apache.poi.hssf.util.HSSFColor$HSSFColorPredefined.<init>(HSSFColor.java:113)
  A  SP 0x00007ffdbdf16060 IP 0x00005db963852335 size=1216  [image code] org.apache.poi.hssf.util.HSSFColor$HSSFColorPredefined.<clinit>(HSSFColor.java:55
  ```
  Spoiler alert, `poi` requires AWT :facepalm:

hubertp commented:
And we may have some issues with `awt` on MacOS, as per https://github.com/oracle/graal/issues/4124

JaroslavTulach commented on std-bits/table/src/main/resources/META-INF/native-image/org/enso/table/resource-config.json:
```diff
@@ -0,0 +1,96 @@
+    "pattern":"\\Qurlhandler.props\\E"
+  }]},
+  "bundles":[{
+    "name":"com.sun.org.apache.xalan.internal.xsltc.compiler.util.ErrorMessages",
```
Xalan! Isn't it generating bytecode on the fly? I remember these used to be [Gregor Samsa class](https://github.com/apache/xalan-java/blob/6fd607c78ebb07ded14521a8ab460a7f4ba4fc1b/xalan/src/main/java/org/apache/xalan/xsltc/compiler/XSLTC.java#L337) which resembled not only transformation but also bytecode generator. How that's gonna work with NI?...

JaroslavTulach commented on test/Table_Tests/src/In_Memory/Table_Spec.enso:
```diff
@@ -188,7 +188,8 @@ add_specs suite_builder =
 
-        group_builder.specify "should work with polyglot values coming from JS" <|
+        pending_js_missing = if Polyglot.is_language_installed "js" then Nothing else "Can't run JavaScript tests, language `js` is not installed."
+        group_builder.specify "should work with polyglot values coming from JS" pending=pending_js_missing <|
```
I see. JavaScript is no longer in the `bin/enso` and thus these tests have to be pending.

JaroslavTulach commented:
> And we may have some issues with awt on MacOS, as per https://github.com/oracle/graal/issues/4124

Yes we may not be the only ones trying to avoid dependency on AWT:
- https://github.com/apache/poi/pull/425
- _"The problem is that AWT is not available in a GraalVM native image built"_: https://github.com/apache/poi/pull/425#issuecomment-1416785554
- there even is a solution: https://github.com/apache/poi/pull/425#issuecomment-1418257194 by Quarkus!?
    - is there a Quarkus version of Apache POI?
    - then we could use that one as it is going to be GraalVM friendly
- FYI: https://github.com/quarkiverse/quarkus-poi

> ```
> java.lang.ClassLoader.loadLibrary(ClassLoader.java:106)
>   i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.lang.Runtime.loadLibrary0(Runtime.java:916)
>   i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.lang.System.loadLibrary(System.java:2063)
>   i  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.awt.Toolkit$2.run(Toolkit.java:1384)
>   A  SP 0x00007ffdbdf15460 IP 0x00005db9626f156f size=16    [image code] java.awt.Toolkit$2.run(Toolkit.java:1382)
>   A  SP 0x00007ffdbdf15470 IP 0x00005db962af378a size=64    [image code] 
>   A  SP 0x00007ffdbdf154e0 IP 0x00005db9626f18a8 size=16    [image code] java.awt.Toolkit.<clinit>(Toolkit.java:1393)
>   i  SP 0x00007ffdbdf16060 IP 0x00005db963852335 size=1216  [image code] org.apache.poi.hssf.util.HSSFColor$HSSFColorPredefined.<init>(HSSFColor.java:113)
>   A  SP 0x00007ffdbdf16060 IP 0x00005db963852335 size=1216  [image code] org.apache.poi.hssf.util.HSSFColor$HSSFColorPredefined.<clinit>(HSSFColor.java:55
> ```
> 
> Spoiler alert, `poi` requires AWT 🤦


Another possible _Quarkiverse_ goodies:
- [OpenCV](https://search.maven.org/search?q=g:io.quarkiverse.opencv)
- [SQLite](https://search.maven.org/artifact/io.quarkiverse.jdbc/quarkus-jdbc-sqlite/3.0.11/jar)
- [Postgres](https://quarkus.io/extensions/io.quarkus/quarkus-jdbc-postgresql/)

Btw. there seems to be some [fancy Quarkus build system](https://github.com/quarkiverse/quarkus-poi/blob/main/deployment/src/main/java/io/quarkiverse/poi/deployment/POIProcessor.java).

JaroslavTulach commented:
I've merged this branch with develop:
```
Merge: a40c4d2bde a1e01fcc51
```
and executed:
```
enso$ ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --run test/Table_Tests
...
5045 tests succeeded.
1 tests failed.
129 tests skipped.
36 groups skipped.
Failed tests: 'should.clear.cache.on.simulated.reload'
```
that's good, isn't it?

hubertp commented:
> that's good, isn't it?

Yes, I resolved all issues yesterday, it seems. Just testing now.

hubertp commented:
> * FYI: https://github.com/quarkiverse/quarkus-poi

Interesting. Now I'm not surprised I had so many issues with it.

<blockquote><img src="https://repository-images.githubusercontent.com/597162470/89d3b974-42b6-4cbf-850b-af798a56f6ed" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/quarkiverse/quarkus-poi">GitHub - quarkiverse/quarkus-poi: Apache POI is an API to access Microsoft Office files. This extension provides integration with Apache POI</a></strong></div><div>Apache POI is an API to access Microsoft Office files. This extension provides integration with Apache POI - quarkiverse/quarkus-poi</div></blockquote>

hubertp committed:
fix moduledeps inconsistency

hubertp committed:
Re-shufflings, de-duplication in resource configs

hubertp committed:
Merge branch 'develop' into wip/hubert/12006-table-tests-ni

hubertp commented:
Re `quarkus-poi` - since this configuration now works, I don't see much pressure to switch. But we _could_ consider it later.

hubertp committed:
legal

jdunkerley reviewed (approved):
Approving from libs side.

JaroslavTulach commented on build.sbt:
```diff
@@ -3718,7 +3720,7 @@ lazy val `engine-runner` = project
         "org.graalvm.nativeimage,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.base,org.graalvm.nativeimage.driver,org.graalvm.nativeimage.librarysupport,org.graalvm.nativeimage.objectfile,org.graalvm.nativeimage.pointsto,com.oracle.graal.graal_enterprise,com.oracle.svm.svm_enterprise"
       val JDK_MODULES =
-        "jdk.localedata,jdk.httpserver,java.naming,java.net.http"
+        "jdk.localedata,jdk.httpserver,java.naming,java.net.http,java.desktop"
```
Including `java.desktop` is undesirable. What can we do to remove it?
- Use quarkus version?
- Replace `java.awt.Color` usage with a substitution?
- Report a bug for later?

hubertp replied:
I would prefer the last option. Configuration is rather fragile and it is easier to improve it incrementally than push it in one go.

JaroslavTulach replied:
Go on, get this in. Otherwise we need to deal https://github.com/enso-org/enso/pull/12073#discussion_r1927401770 forever. Once this change is in, we can polish it more.

hubertp replied:
`quarkus-poi` doesn't work out-of-the-box. It seems it relies on the fact that one builds it for redhat base image:
https://github.com/quarkiverse/quarkus-poi?tab=readme-ov-file#docker

JaroslavTulach commented on engine/language-server/src/main/resources/META-INF/native-image/org/enso/languageserver/reflect-config.json:
```diff
@@ -3197,405 +2879,6 @@
-  "name":"org.enso.table.parsing.IdentityParser"
-},
-{
-  "name":"org.enso.table.parsing.NumberParser"
```
Good. `languageserver` shouldn't care about `std-table` classes.

JaroslavTulach commented on std-bits/base/src/main/resources/META-INF/native-image/org/enso/base/reflect-config.json:
```diff
@@ -0,0 +1,325 @@
+        "name": "org.enso.base.time.FormatterKind"
+    },
+    {
+        "name": "org.enso.base.time.Time_Of_Day_Utils"
```
Classes from `org.enso.base` package belong to `std-base`. Good.

JaroslavTulach commented on std-bits/table/src/main/resources/META-INF/native-image/org/enso/table/jni-config.json:
```diff
@@ -0,0 +1,29 @@
+    "fields":[{"name":"value"}]
+  },
+  {
+    "name":"java.awt.GraphicsEnvironment",
```
It would be better to create substitutions for these methods. Something like
```java
@TargetClass(java.awt.GraphicsEnvironment.class)
class Target_GraphicsEnvironment {
   @Substitution
   public boolean isHeadless() { return true; }
}
```
rather than enabling `java.desktop`.

JaroslavTulach commented on std-bits/table/src/main/resources/META-INF/native-image/org/enso/table/reflect-config.json:
```diff
@@ -0,0 +1,1912 @@
+        "name": "org.enso.table.data.column.operation.cast.CastProblemAggregator"
+    },
+    {
+        "name": "org.enso.table.data.column.operation.cast.ConversionFailure"
```
I would expect these classes to be listed somewhere in `polyglot java import`... Are then not? Then we solved this situation with [Extra_Imports](https://github.com/enso-org/enso/blob/7029c4ad70d1e85344412d69ce85d4ef2abb35f3/distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Extra_Imports.enso#L4) to make them more approachable to libs guys.  Listing them here does the same thing, but I don't think libs guys will ever edit this file.

hubertp replied:
Many of these have been generated by the agent.

JaroslavTulach commented on std-bits/table/src/main/resources/META-INF/native-image/org/enso/table/reflect-config.json:
```diff
@@ -0,0 +1,1912 @@
+        ]
+    },
+    {
+        "name": "org.openxmlformats.schemas.spreadsheetml.x2006.main.impl.WorksheetDocumentImpl",
```
This is pretty log file!

JaroslavTulach commented:
> Re `quarkus-poi` - since this configuration now works, I don't see much pressure to switch. But we _could_ consider it later.

The added dependency on `java.desktop` would be my biggest motivation to switch. But yes, it can be done later.

JaroslavTulach reviewed (approved):
- great you got it working
- I don't like the new dependency on `java.desktop`
- it its a shame for server side software to depend on AWT...
- but let's merge while it work and worry about architecture purity later

JaroslavTulach commented:
- please add the `test/Table_Tests` to the CI run. Right now it has to be done manually here:

https://github.com/enso-org/enso/blob/8cc8c4f7b0919ce583c5ac557aac9a2e9236e980/build_tools/build/src/engine/context.rs#L684

- this all will change when #12123 is implemented

hubertp committed:
Remove xalan bundle

hubertp committed:
PR review

hubertp committed:
nit

hubertp commented:
It appears to fail now with some weird errors that I can't reproduce locally:
```
2025-01-24T12:38:11.4550135Z  INFO ide_ci::program::command: enso ℹ️     - [PENDING] date_diff and date_add should work correctly with DST
2025-01-24T12:38:11.4550795Z  INFO ide_ci::program::command: enso ℹ️         Reason: May be failing on some Database configurations. ToDo: investigate - https://github.com/enso-org/enso/issues/7326
2025-01-24T12:38:14.9047076Z  INFO ide_ci::program::command: enso ℹ️ [In-Memory] Table.filter: [28/28, 1004ms]
2025-01-24T12:38:14.9049598Z  INFO ide_ci::program::command: enso ℹ️     - [PENDING] text operations should support Unicode normalization (like)
2025-01-24T12:38:14.9050446Z  INFO ide_ci::program::command: enso ℹ️         Reason: There is a bug with Java Regex in Unicode normalized mode (CANON_EQ) with quoting.
2025-01-24T12:38:14.9051254Z  INFO ide_ci::program::command: enso ℹ️ https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8032926
2025-01-24T12:38:15.3297855Z  INFO ide_ci::program::command: enso ℹ️ [WARN] [2025-01-24T13:38:15+01:00] [org.apache.poi.POIDocument] DocumentSummaryInformation property set came back as null
2025-01-24T12:38:15.3298559Z  INFO ide_ci::program::command: enso ℹ️ [WARN] [2025-01-24T13:38:15+01:00] [org.apache.poi.POIDocument] SummaryInformation property set came back as null
2025-01-24T12:38:15.5885932Z  INFO ide_ci::program::command: enso ⚠️ AAAo
2025-01-24T12:38:15.5886969Z  INFO ide_ci::program::command: enso ℹ️ AAAo
2025-01-24T12:38:15.5887411Z  INFO ide_ci::program::command: enso ⚠️ java.lang.Exception
2025-01-24T12:38:15.5888098Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.table.excel.ExcelConnectionPool.release(ExcelConnectionPool.java:206)
2025-01-24T12:38:15.5888945Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.table.excel.ReadOnlyExcelConnection.close(ReadOnlyExcelConnection.java:26)
2025-01-24T12:38:15.5889745Z  INFO ide_ci::program::command: enso ⚠️ 	at java.base@21.0.2/java.lang.reflect.Method.invoke(Method.java:580)
2025-01-24T12:38:15.5890741Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostMethodDesc$SingleMethod$MethodReflectImpl.reflectInvoke(HostMethodDesc.java:305)
2025-01-24T12:38:15.5891880Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostMethodDesc$SingleMethod$MethodReflectImpl.invoke(HostMethodDesc.java:290)
2025-01-24T12:38:15.5893080Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.GuestToHostCodeCache$GuestToHostInvokeReflect.executeImpl(GuestToHostCodeCache.java:110)
2025-01-24T12:38:15.5894387Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.GuestToHostRootNode.execute(GuestToHostRootNode.java:80)
2025-01-24T12:38:15.5895422Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:745)
2025-01-24T12:38:15.5896623Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callInlined(OptimizedCallTarget.java:550)
2025-01-24T12:38:15.5897945Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedRuntimeSupport.callInlined(OptimizedRuntimeSupport.java:250)
2025-01-24T12:38:15.5898967Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.GuestToHostRootNode.guestToHostCall(GuestToHostRootNode.java:102)
2025-01-24T12:38:15.5900169Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostMethodDesc$SingleMethod$ReflectBase.invokeGuestToHost(HostMethodDesc.java:259)
2025-01-24T12:38:15.5901152Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNode.doInvoke(HostExecuteNode.java:877)
2025-01-24T12:38:15.5902017Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNode.doFixed(HostExecuteNode.java:140)
2025-01-24T12:38:15.5902975Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNodeGen$Inlined.executeAndSpecialize(HostExecuteNodeGen.java:403)
2025-01-24T12:38:15.5904011Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNodeGen$Inlined.execute(HostExecuteNodeGen.java:363)
2025-01-24T12:38:15.5904988Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostObject.invokeMember(HostObject.java:465)
2025-01-24T12:38:15.5905986Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached.invokeMemberNode_AndSpecialize(HostObjectGen.java:7156)
2025-01-24T12:38:15.5907143Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached.invokeMember(HostObjectGen.java:7142)
2025-01-24T12:38:15.5908746Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibrary$Asserts.invokeMember(InteropLibrary.java:3595)
2025-01-24T12:38:15.5909800Z  INFO ide_ci::program::command: enso ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch.invokeMember(InteropLibraryGen.java:8549)
2025-01-24T12:38:15.5910907Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.resolver.HostMethodCallNode.resolveHostMethod(HostMethodCallNode.java:223)
2025-01-24T12:38:15.5912286Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.resolver.HostMethodCallNodeGen.executeAndSpecialize(HostMethodCallNodeGen.java:157)
2025-01-24T12:38:15.5913435Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.resolver.HostMethodCallNodeGen.execute(HostMethodCallNodeGen.java:119)
2025-01-24T12:38:15.5914437Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.InvokeMethodNode.doPolyglot(InvokeMethodNode.java:546)
2025-01-24T12:38:15.5915467Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.InvokeMethodNodeGen.executeAndSpecialize(InvokeMethodNodeGen.java:878)
2025-01-24T12:38:15.5916470Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.InvokeMethodNodeGen.execute(InvokeMethodNodeGen.java:535)
2025-01-24T12:38:15.5917413Z  INFO ide_ci::program::command: enso ⚠️ 	at org.enso.interpreter.node.callable.InvokeCallableNode.invokeDynamicSymbol(InvokeCallableNode.java:276)
```

hubertp committed:
Update table's reflect config

CI fails with a useless `java.lang.Exception` which is not reproducible
locally. Attempting to blindly add a few methods/fields, hoping that it
will fix it.

radeusgd commented:
> It appears to fail now with some weird errors that I can't reproduce locally:

I think the huge stack trace you see is not an error, it is just a log message that was leftover. This is getting fixed in #12139.

hubertp committed:
Potentially fix Windows NI failures

hubertp commented:
> > It appears to fail now with some weird errors that I can't reproduce locally:
> 
> I think the huge stack trace you see is not an error, it is just a log message that was leftover. This is getting fixed in #12139.

You are probably right, yes. The stacktraces were so common and so big that it hid issues with non-Linux OS.

hubertp commented:
And, as suspected, I'm hitting the awt problem on MacOS.

hubertp committed:
Workaround java.awt dependency via substitution

java.awt support on MacOS is still missing therefore
bundling/configuring resources for it only makes sense on Linux/Windows.
That was unsatisafactory.

This change uses native image's substitution mechanism to avoid any awt
library loading. The latter will not be functional when one attempts to
use it but our implementation does not rely on it at all.

hubertp commented:
I used substitution after all because of [lack of support](https://github.com/oracle/graal/issues/4124) for awt on MacOS - our builds were failing.This is acceptable as our implementation does not rely on that part at all.

JaroslavTulach commented on std-bits/table/src/main/java/org/enso/nativeimage/workarounds/ReplacementGraphicsEnvironment.java:
```diff
@@ -0,0 +1,10 @@
+@TargetClass(java.awt.GraphicsEnvironment.class)
+final class ReplacementGraphicsEnvironment {
+    @Substitute
+    public static boolean isHeadless() { return true; }
```
Yup, this one is straightforward.

JaroslavTulach commented on std-bits/table/src/main/java/org/enso/nativeimage/workarounds/ReplacementToolkit.java:
```diff
@@ -0,0 +1,17 @@
+import com.oracle.svm.core.annotate.Substitute;
+
+@TargetClass(java.awt.Toolkit.class)
+final class ReplacementToolkit {
```
Is that all to get rid of dependency on AWT?

hubertp replied:
For our use-cases, yes. Note that this renders AWT pretty much useless but that's OK. I suspect NI eliminates most if not all java.desktop code. `java.beans` (in [DataFormatter](https://github.com/apache/poi/blob/trunk/poi/src/main/java/org/apache/poi/ss/usermodel/DataFormatter.java)) still requires `java.desktop` and I couldn't get rid of that dependency the same way; this time because of `log4j` :facepalm: .

JaroslavTulach reviewed (approved):
Shouldn't you also remove dependency on `java.desktop` module?

hubertp committed:
fmt

hubertp commented:
> Shouldn't you also remove dependency on `java.desktop` module?

No. It needs it to perform the substitution at build-time, sadly.
